python/ray/tests/test_runtime_env_2.py

```diff
@@ -125,6 +130,105 @@ def run(runtime_env):
+                    assert user_info not in line, (file, user_info, line)
+
+
+class TestNoUserInfoInLogs:
+    """Test that no user info (e.g. runtime env env vars) show up in the logs."""
+
+    def test_assert_no_user_info_in_logs(self, shutdown_only):
+        """Test assert_no_user_info_in_logs does not spuriously pass."""
+        ray.init()
+        with pytest.raises(AssertionError):
+            assert_no_user_info_in_logs("ray")
+        assert_no_user_info_in_logs("ray", file_whitelist=["*"])
+
+    def test_basic(self, start_cluster, monkeypatch, tmp_path, shutdown_only):
+        """Test driver with and without Ray Client."""
+
+        cluster, address = start_cluster
+
+        # Runtime env logs may still appear in debug logs. Check the debug flag is off.
+        assert os.getenv("RAY_BACKEND_LOG_LEVEL") != "debug"
+
+        # Reuse the same "secret" for working_dir, pip, env_vars for convenience.
+        USER_SECRET = "pip-install-test"
+        working_dir = tmp_path / USER_SECRET
+        working_dir.mkdir()
+        runtime_env = {
+            "working_dir": str(working_dir),
+            "pip": [USER_SECRET],
+            # Append address to ensure different runtime envs for client and non-client
+            # code paths to force reinstalling the runtime env instead of reusing it.
+            "env_vars": {USER_SECRET: USER_SECRET + str(address)},
+        }
+        ray.init(runtime_env=runtime_env)
+
+        # Run a function to ensure the runtime env is set up.
+        @ray.remote
+        def f():
+            return os.environ.get(USER_SECRET)
+
+        assert USER_SECRET in ray.get(f.remote())
+
+        @ray.remote
+        class Foo:
+            def __init__(self):
+                self.x = os.environ.get(USER_SECRET)
+
+            def get_x(self):
+                return self.x
+
+        foo = Foo.remote()
+        assert USER_SECRET in ray.get(foo.get_x.remote())
+
+        # Generate runtime env failure logs too.
+        bad_runtime_env = {
+            "pip": ["pkg-which-sadly-does-not-exist"],
+            "env_vars": {USER_SECRET: USER_SECRET},
+        }
+        with pytest.raises(Exception):
+            ray.get(f.options(runtime_env=bad_runtime_env).remote())
+        with pytest.raises(Exception):
+            foo2 = Foo.options(runtime_env=bad_runtime_env).remote()
+            ray.get(foo2.get_x.remote())
+
+        using_ray_client = address.startswith("ray://")
+
+        # Test Ray Jobs API codepath. Skip for ray_minimal because Ray Jobs API
+        # requires ray[default].
+        if not using_ray_client and os.environ.get("RAY_MINIMAL") != "1":
+            client = JobSubmissionClient()
+            client.submit_job(entrypoint="echo 'hello world'", runtime_env=runtime_env)
+            client.submit_job(
+                entrypoint="echo 'hello world'", runtime_env=bad_runtime_env
+            )
+
+        # Allow time for log files to be written.
+        time.sleep(5)
```

Comment:
how about wait until the job finishes instead of sleeping? that should avoid the sleep